### PR TITLE
feat: [#183443674] update c-corp formation

### PR DESCRIPTION
**Pull request by jphechter**
NOTE: Waiting on feedback from the PrepareFiling API team. Currently, cannot successfully submit. Getting the following error:

```
[{"Valid":false,"ErrorMessage":"Required","Name":"Formation.ForeignGoodStandingFile"}]
```

---

## Description

Updates the flow for foreign c-corp formation.

### Ticket

[183443674](https://www.pivotaltracker.com/story/show/183443674)

### Approach

- Create new components associated with new formation questions
  - `PracticesLaw`
  - `ForeignCertificate`
- Create a new `FileInput` component using the USWDS component and functionality as a launchpad
- Established new types and updated config files as necessary
- Updated existing tests to be more flexible and to accommodate for the new `foreignGoodStandingFile` that will be passed around alongside formation data

We are not storing the `foreignGoodStandingFile` but it needs to be maintained in state while the user is filling out the form prior to submission. To accomplish this, we created a new state object. Passing this around required updating a number of functions to accept a props object (easier to handle optional props) and modifying some types.

### Steps to Test

You can experiment with this work by creating a foreign c-corp. The UI changes should be on the formation Business step and Review step. Clicking submit should hit the NICUSA API. Check for errors and successful submission to NICUSA.

### Notes

n/a

## Code author checklist

- [X] I have performed a self-review of my code
- [X] I have created and/or updated relevant documentation, if necessary
- [ ] I have not used any relative imports
  - This does in fact use a relative import to get the USWDS library to play nicely. 
- [X] I have checked for and removed instances of unused config from CMS
- [X] I have pruned any instances of unused code

### CODE REVIEW

File: api/src/client/ApiFormationClient.test.ts [hunk 1]
```diff
@@ -567,13 +568,22 @@ describe("ApiFormationClient", () => {
           formationFormData
         );
 
-        await client.form(userData, "hostname.com/form-business");
+        const foreignGoodStandingFile = generateInputFile({
+          fileType: "PNG",
+          base64Contents: "this is my string",
+        });
```
Comment (aloverso) [inline api/src/client/ApiFormationClient.test.ts:574]: I don't think we need to specify the fileType and base64Contents? We're never testing for them later on. I tend to push for reducing/removing any extraneous info from tests so that everything that is present has a reason to be. I think here we could just let the random generator do its work, we don't care the exact values.

File: api/src/client/ApiFormationClient.test.ts [hunk 2]
```diff
@@ -593,7 +603,7 @@ describe("ApiFormationClient", () => {
 
             BusinessInformation: {
               CompanyOrigin: "Foreign",
-              PracticesLaw: false,
+              PracticesLaw: "Yes",
```
Comment (aloverso) [inline api/src/client/ApiFormationClient.test.ts:606]: Why do we expect this to be "Yes"? I don't see that anywhere in the setup. Shouldn't this come from setting it to `true` somewhere in the form data during setup?

File: api/src/client/ApiFormationClient.ts [hunk 3]
```diff
@@ -251,7 +265,7 @@ export const ApiFormationClient = (config: ApiConfig, logger: LogWriterType): Fo
           BusinessName: formationFormData.businessName,
           BusinessDesignator: formationFormData.businessSuffix as Exclude<BusinessSuffix, undefined>,
           Naic: naicsCode,
-          PracticesLaw: isForeignCorp ? false : undefined,
+          PracticesLaw: isForeignCorp ? "Yes" : undefined,
```
Comment (aloverso) [inline api/src/client/ApiFormationClient.ts:268]: Re my comment in the test file. Looks like we hardcoded this to "Yes" temporarily and never came back to actually set this value based on the formationFormData. This should be updated

File: shared/src/formationData.ts [hunk 4]
```diff
@@ -145,10 +142,11 @@ export interface FormationFormData extends FormationAddress {
   readonly contactPhoneNumber: string;
   readonly foreignStateOfFormation: StateNames | undefined;
   readonly foreignDateOfFormation: string | undefined; // YYYY-MM-DD
-  readonly foreignGoodStandingFile: ForeignGoodStandingFileObject | undefined;
+  readonly willPracticeLaw: "Yes" | "No" | undefined;
```
Comment (aloverso) [inline shared/src/formationData.ts:145]: why are we redefining this when we have the value up on line 101?
Also, I would argue we don't want to store this as "Yes" and "No" but rather as a boolean true/false. The Yes/No is how the API accepts it but since this is our domain layer, we have the freedom to choose a more appropriate representation for our type. And being able to write things like `if (willPracticeLaw)` is more in line with how we expect things to work, rather than having to do `if (willPracticeLaw === "Yes")`

File: shared/tsconfig.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "extends": "./tsconfig.paths.json",
   "compilerOptions": {
     "esModuleInterop": true,
-    "lib": ["ESNext"],
+    "lib": ["ESNext", "DOM"],
```
Comment (aloverso) [inline shared/tsconfig.json:5]: What is this for?

File: web/src/components/njwds-extended/FileInput.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+  PNG: "image/png",
+};
+
+const InputStringToFileType: Record<string, AcceptedFileType> = (
+  Object.keys(FileTypeToInputString) as AcceptedFileType[]
+).reduce((acc, key) => {
+  const value = FileTypeToInputString[key];
+  acc[value] = key;
+  return acc;
+}, {} as Record<string, AcceptedFileType>);
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:26]: Might be nice to extract this into a generic "flip dictionary" helper method

File: web/src/components/njwds-extended/FileInput.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+export const FileInput = (props: Props): ReactElement => {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const fileInput = require("../../../../node_modules/@uswds/uswds/packages/usa-file-input/src");
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:30]: a) Is this really how the USWDS wants us to use this component?
b) Why does it need to be here instead of a regular top-level import?

File: web/src/components/njwds-extended/FileInput.tsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  const fileInputRef = useRef<HTMLInputElement>(null);
+  const BYTES_IN_A_MB = 1048576;
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:32]: could go outside the function so it's not constantly redefined. Small tweak

File: web/src/components/njwds-extended/FileInput.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    // Prevent the user from selecting a file that is too large or of the wrong file type
+    const preventInvalidFileSelection = (e: Event): void => {
+      const eventTarget = e.target as HTMLInputElement;
+      if (eventTarget.files && eventTarget.files.length > 0) {
+        const selectedFile = eventTarget.files[0];
+        if (props.accept && !props.accept.includes(InputStringToFileType[selectedFile.type])) {
+          e.stopImmediatePropagation();
+          setFileSizeError({
+            hasError: true,
+            errorMessage: `${selectedFile.name} file format is not supported.`,
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:78]: I think this error message (and the one below it) should be externalized into the config files. They can still use the `${}` string interpolation. There's a helper called `templateEval` that allows us to do that with config strings

File: web/src/components/njwds-extended/FileInput.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    // Event listener is added to the file input element ahead of USWDS listeners - React onChange will be added after
+    fileInputElement?.addEventListener("change", (e: Event) => {
+      preventInvalidFileSelection(e);
+    });
+    if (typeof fileInput.on === "function" && typeof fileInput.off === "function") {
+      fileInput.on(fileInputElement); // initialize USWDS fileInput component
+    }
+  }, [fileInput, props.accept, size, props.maxSizeInMegabytes]);
+
+  useEffect(() => {
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:99]: because this is a weird useEffect, it might be nice to use the named-function approach to a useEffect here, so that we can give it a function name to help make it more readable & explain what's going on for people reading it

File: web/src/components/njwds-extended/FileInput.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+export const FileInput = (props: Props): ReactElement => {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const fileInput = require("../../../../node_modules/@uswds/uswds/packages/usa-file-input/src");
+  const fileInputRef = useRef<HTMLInputElement>(null);
+  const BYTES_IN_A_MB = 1048576;
+  const size = props.maxSizeInMegabytes ? props.maxSizeInMegabytes * BYTES_IN_A_MB : undefined;
+
+  const [hasFileBeenRead, setHasFileBeenRead] = useState<boolean>(false);
+  const [fileSizeError, setFileSizeError] = useState({ hasError: false, errorMessage: "" });
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:36]: It seems a little misleading to call this a `fileSizeError` when it also represents a file type error as well. Maybe just `fileUploadError` or something similar?

File: web/src/components/tasks/business-formation/business/MainBusiness.tsx [hunk 12]
```diff
@@ -45,26 +47,34 @@ export const MainBusiness = (): ReactElement => {
...
+              <FormationDate fieldName="foreignDateOfFormation" />
+            </WithErrorBar>
           </WithErrorBar>
-        </WithErrorBar>
+          {state.formationFormData.legalType === "foreign-c-corporation" && (
+            <PracticesLaw hasError={doesFieldHaveError("willPracticeLaw")} />
+          )}
+          {state.formationFormData.legalType === "foreign-c-corporation" && (
+            <ForeignCertificate hasError={doesFieldHaveError("foreignGoodStandingFile")} />
+          )}
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/business/MainBusiness.tsx:76]: These could both go within the same logic check instead of repeating it?

Comment (aloverso) [reply web/src/components/tasks/business-formation/business/MainBusiness.tsx:76]: Also - I'm noticing that the other components like `FormationDate` and `ForeignStateOfFormation` are independent components, and then are wrapped in the WithErrorBar here in the parent-display layout. Whereas you included the WithErrorBar wrapper internal to the ForeignCertificate. 

I'd argue that makes more sense to keep it external because the error bar relies on assumptions about layout. Specifically the MOBILE/DESKTOP view. If the file input were side-by-side with another component, the chosen "ALWAYS" prop would have to change, and that would be hard to track internally. Hence my argument for keeping the ErrorBar at the top level like we do for the other components here.

File: web/src/components/tasks/business-formation/business/PracticesLaw.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        willPracticeLaw: value,
+      };
+    });
+  };
+
+  return (
+    <WithErrorBar hasError={hasError} type="ALWAYS">
+      <FormControl variant="outlined" fullWidth className="padding-bottom-2">
+        <RadioGroup
+          aria-label={"Foreign address type"}
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/business/PracticesLaw.tsx:30]: copy paste error? Usually we use `camelCaseToSentenceCase` of the field name for aria labels

File: web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    <>
+      <ReviewSectionHeader
+        header="Foreign Certificate of Good Standing"
+        stepName="Business"
+        testId="foreign-certificate"
+      />
+      {props.foreignGoodStandingFile?.filename ? (
+        <div style={{ display: "flex", alignItems: "center" }}>
+          <img
+            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx:22]: what is this?

File: web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+interface Props {
+  foreignGoodStandingFile: InputFile | undefined;
+}
+
+export const ReviewForeignCertificate = (props: Props): ReactElement => {
+  const { Config } = useConfig();
+  return (
+    <>
+      <ReviewSectionHeader
+        header="Foreign Certificate of Good Standing"
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx:15]: should be based on the config label?

File: web/src/components/tasks/business-formation/review/ReviewStep.tsx [hunk 16]
```diff
@@ -30,6 +32,12 @@ export const ReviewStep = (): ReactElement => {
       <div data-testid="review-step">
         <BusinessNameAndLegalStructure isReviewStep />
         <ReviewBusinessSuffixAndStartDate />
+        {isForeignCCorp && (
+          <>
+            <ReviewText header="Will you practice law?" fieldName="willPracticeLaw" stepName="Business" />
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/review/ReviewStep.tsx:37]: should be based on the config label?

File: web/src/lib/api-client/apiClient.ts [hunk 17]
```diff
@@ -30,8 +31,13 @@ export const checkLicenseStatus = (nameAndAddress: NameAndAddress): Promise<User
   return post(`/license-status`, nameAndAddress);
 };
 
-export const postBusinessFormation = (userData: UserData, returnUrl: string): Promise<UserData> => {
-  return post(`/formation`, { userData, returnUrl });
+// Pass file obj here
```
Comment (aloverso) [inline web/src/lib/api-client/apiClient.ts:34]: delete comment

File: web/src/styles/base/global.scss [hunk 18]
```diff
@@ -209,6 +209,10 @@ ol {
   }
 }
 
+.usa-file-input {
+  max-width: unset;
+}
```
Comment (aloverso) [inline web/src/styles/base/global.scss:214]: Is this overriding the default? If so, did you check for sure it works / is actually overridden in browser? usually in order to to override an njwds setting, we have to use `!important`. I'm surprised to see this would work like this

File: web/test/helpers/helpers-formation.tsx [hunk 19]
```diff
@@ -180,6 +180,8 @@ export type FormationPageHelpers = {
   fillAndSubmitAddressModal: (overrides: Partial<FormationSignedAddress>, fieldName: string) => Promise<void>;
   clickSubmit: () => Promise<void>;
   selectDate: (value: DateObject, fieldType: "Business start date" | "Foreign date of formation") => void;
+  uploadFile: (file: File) => void;
+  willPracticeLaw: (response?: boolean) => void;
```
Comment (aloverso) [inline web/test/helpers/helpers-formation.tsx:184]: I think it'd be more readable if this function was named something that clearly described the action being performed.

File: api/src/client/ApiFormationClient.test.ts (hunk 1 above)
Comment (jphechter) [reply api/src/client/ApiFormationClient.test.ts:574]: Fair point. Included!

File: api/src/client/ApiFormationClient.ts (hunk 3 above)
Comment (jphechter) [reply api/src/client/ApiFormationClient.ts:268]: 👀

File: shared/tsconfig.json (hunk 5 above)
Comment (jphechter) [reply shared/tsconfig.json:5]: I had added that to use the `File` type defined by React. I don't think I need this any more.

File: web/src/styles/base/global.scss (hunk 18 above)
Comment (jphechter) [reply web/src/styles/base/global.scss:214]: Yup, this is overriding the default value. This is pulling styles from the USWDS package. I'm not sure how NJWDS implements USWDS.
Screenshot 2023-04-11 at 10 05 07 AM

File: web/src/components/njwds-extended/FileInput.tsx (hunk 11 above)
Comment (jphechter) [reply web/src/components/njwds-extended/FileInput.tsx:36]: included

File: web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx (hunk 14 above)
Comment (jphechter) [reply web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx:22]: This is the preview image that's used by the USWDS file-input component. I used it here for continuity. I could extrapolate this to a global variable or reference it some other way if you'd like.

File: web/src/components/tasks/business-formation/business/MainBusiness.tsx (hunk 12 above)
Comment (jphechter) [reply web/src/components/tasks/business-formation/business/MainBusiness.tsx:76]: That reasoning makes sense, but I'm leaving `WithErrorBar` nested. I see this pattern in other components like `BusinessFormationTextField`. These components really only exist to be a part of this form and aren't affected by the surrounding items. I think it just eliminates a little clutter from `MainBusiness`.

File: web/src/components/njwds-extended/FileInput.tsx (hunk 6 above)
Comment (jphechter) [reply web/src/components/njwds-extended/FileInput.tsx:26]: included

File: web/src/components/njwds-extended/FileInput.tsx (hunk 10 above)
Comment (jphechter) [reply web/src/components/njwds-extended/FileInput.tsx:99]: included

File: api/src/client/ApiFormationClient.test.ts [hunk 20]
```diff
@@ -593,11 +600,11 @@ describe("ApiFormationClient", () => {
 
             BusinessInformation: {
               CompanyOrigin: "Foreign",
-              PracticesLaw: false,
               Business: "ForeignForProfitCorporation",
               BusinessName: formationFormData.businessName,
               BusinessDesignator: formationFormData.businessSuffix,
               Naic: userData.profileData.naicsCode,
+              PracticesLaw: "No",
```
Comment (aloverso) [inline api/src/client/ApiFormationClient.test.ts:607]: Same comment as before - shouldn't this be driven by `willPracticeLaw` being set to `false` somewhere in the setup section of this test?

File: content/src/fieldConfig/business-formation.json [hunk 21]
```diff
@@ -205,6 +205,15 @@
...
         "error": "Select a State"
       },
+      "foreignGoodStandingFile": {
+        "label": "Foreign Certificate of Good Standing",
+        "helperText": "Select a PNG or PDF file. Max file size: 3MB.",
+        "errors": {
+          "fileSize": "${fileName} exceeds maximum size of ${maxFileSize} MB.",
+          "fileType": "${fileName} file format is not supported.",
+          "required": "A Certificate of Good Standing is required."
+        }
```
Comment (aloverso) [inline content/src/fieldConfig/business-formation.json:215]: Not a huge thing, but something for consideration: sometimes nesting in the CMS ui is helpful to organize things and close extraneous information. But the more nesting we do, the more grey bars and boxes that Netlify creates on the screen. 

So I tend to nest things when it is clear that you'd want to look at one thing while being able to collapse unrelated info. I would argue that a content team member looking at the Foreign Good Standing File would want to see all the fields, and there's not much value in being able to see label & helpertext but collapse the error fields. So at that point, the extra boxes and nesting in the CMS UI is more visually distracting than helpful. If you haven't looked at the CMS to see what I mean, I recommend it. Just a thoughtt.

File: web/src/components/njwds-extended/FileInput.test.tsx [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+        value={undefined}
+      />
+    );
+
+    const uploader = screen.getByTestId("file-input");
+
+    fireEvent.change(uploader, { target: { files: [file] } });
+
+    expect(mockOnChange).not.toHaveBeenCalled();
+    expect(screen.getByText("cool.png file format is not supported.")).toBeInTheDocument();
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.test.tsx:83]: If this is in config, we should use the config in this. If content changes the message, then this test will start failing

Comment (aloverso) [reply web/src/components/njwds-extended/FileInput.test.tsx:83]: and the one above

File: web/src/components/tasks/business-formation/business/PracticesLaw.tsx [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+    setFormationFormData((previousState) => {
+      return {
+        ...previousState,
+        willPracticeLaw,
+      };
+    });
+  };
+
+  return (
+    <WithErrorBar hasError={hasError} type="ALWAYS">
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/business/PracticesLaw.tsx:28]: Same comment about withErrorBar as the other component

File: web/src/components/njwds-extended/FileInput.test.tsx (hunk 22 above)
Comment (jphechter) [reply web/src/components/njwds-extended/FileInput.test.tsx:83]: I provided a default value to the `errorMessageFileSize` and `errorMessageFileType` values because these errors require multiple currently optional params. My thinking was that I was trying to test the defaults which is why I didn't pull from the Config, but that's really not what this test is currently doing.

Maybe I'm over-engineering this. It might not be entirely obvious where the default provided copy is coming from. I was trying to account for an edge case that might not really be a big deal.

File: api/src/client/ApiFormationClient.test.ts (hunk 20 above)
Comment (jphechter) [reply api/src/client/ApiFormationClient.test.ts:607]: Fixed! Thank you.

File: content/src/fieldConfig/business-formation.json (hunk 21 above)
Comment (jphechter) [reply content/src/fieldConfig/business-formation.json:215]: I haven't poked a ton around in there. I'll de-structure this.

File: web/src/components/njwds-extended/FileInput.tsx [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+import { ReactElement, useEffect, useLayoutEffect, useRef, useState } from "react";
+
+interface Props {
+  accept?: AcceptedFileType[];
+  errorMessageFileSize?: string;
+  errorMessageFileType?: string;
+  errorMessageRequired: string;
+  hasError?: boolean;
+  helperText: string;
+  maxSizeInMegabytes?: number; // filesize in bytes (1,048,576 = 1 MB)
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:12]: comment is incorrect, should be deleted

File: web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx (hunk 14 above)
Comment (aloverso) [reply web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx:22]: lets give it a variable name so other people can know this also

File: web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+  const { Config } = useConfig();
+  return (
+    <ReviewSection
+      buttonText={Config.formation.general.editButtonText}
+      header={Config.formation.fields.foreignGoodStandingFile.label}
+      stepName="Business"
+      testId="edit-foreign-certificate-step"
+    >
+      {props.foreignGoodStandingFile?.filename ? (
+        <div style={{ display: "flex", alignItems: "center" }}>
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx:20]: classname "fac" in flex.scss

File: web/src/components/tasks/business-formation/business/PracticesLaw.tsx [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+interface Props {
+  hasError: boolean;
+}
+
+export const PracticesLaw = (props: Props): ReactElement => {
+  const { hasError } = props;
+  const { Config } = useConfig();
+  const { state, setFormationFormData } = useContext(BusinessFormationContext);
+
+  const onChange = (value: "true" | "false"): void => {
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/business/PracticesLaw.tsx:16]: Just noticed this - can we just use a boolean instead of string values?

File: api/src/client/ApiFormationClient.test.ts [hunk 27]
```diff
@@ -665,6 +671,92 @@ describe("ApiFormationClient", () => {
...
+      describe("getPracticesLaw", () => {
+        let stubResponse: ApiResponse;
+
+        beforeEach(() => {
+          jest.resetAllMocks();
+          stubResponse = generateApiResponse({});
+          mockAxios.post.mockResolvedValue({ data: stubResponse });
+        });
+
+        it("PracticesLaw is undefined if not provided", async () => {
```
Comment (aloverso) [inline api/src/client/ApiFormationClient.test.ts:684]: not a big deal but just a note that we try to write test names as a sentence starting with "it..."

Comment (aloverso) [reply api/src/client/ApiFormationClient.test.ts:684]: The reason for this is that it's a technique to encourage behavioral testing. As a principle, we always want to be testing some _output_ or behavior of the thing under test. So if we can describe what "it" does - that's its behavior. And if there's something that we can't put in this format, it's probably not actually observable behavior and we should consider how to refactor the test

File: web/src/components/njwds-extended/FileInput.test.tsx [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+        filename: "cool.png",
+      });
+    });
+  });
+
+  it("refuses files larger than the provided max file size", async () => {
+    const mockOnChange = jest.fn();
+    const errorMessageFileSize = Config.formation.fields.foreignGoodStandingFile.errorMessageFileSize;
+
+    file = new File(["x".repeat(1048576 / 2)], "cool.png", { type: "image/png" }); // .5 MB file
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.test.tsx:44]: this would be an example where I recommend a variable named `const halfMegabyteFile = "x".repeat...` instead of using a comment, to have the code be "self-documenting"

File: web/src/components/njwds-extended/FileInput.test.tsx (hunk 22 above)
Comment (aloverso) [reply web/src/components/njwds-extended/FileInput.test.tsx:86]: Is this comment still relevant? We want to test that this value came from config, right? Since there's no default anymore?

File: web/src/components/njwds-extended/FileInput.tsx [hunk 29]
```diff
@@ -0,0 +1,170 @@
...
+        setFileUploadError({ hasError: false, errorMessage: "" });
+      };
+      reader.readAsDataURL(uploadedFile);
+    }
+  };
+
+  useLayoutEffect(() => {
+    const fileInputElement = fileInputRef.current;
+
+    // Prevent the user from selecting a file that is too large or of the wrong file type
```
Comment (aloverso) [inline web/src/components/njwds-extended/FileInput.tsx:88]: No need for this comment because the function name is self-evident

File: web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx [hunk 30]
```diff
@@ -0,0 +1,35 @@
+import { useConfig } from "@/lib/data-hooks/useConfig";
+import { InputFile } from "@businessnjgovnavigator/shared/formationData";
+import { ReactElement } from "rehype-react/lib";
+import { ReviewSection } from "./section/ReviewSection";
+
+interface Props {
+  foreignGoodStandingFile: InputFile | undefined;
+}
+
+const filePreviewImage = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"; // taken from USWDS
```
Comment (aloverso) [inline web/src/components/tasks/business-formation/review/ReviewForeignCertificate.tsx:10]: could be called `uswdsFileImagePreview` to remove the need for a comment

**Comment by jphechter** [comment]
Rebased, but will follow up on comments shortly.
